src/main/java/org/kuali/kfs/module/ar/document/service/impl/ContractsGrantsInvoiceDocumentServiceImpl.java

```diff
@@ -0,0 +1,2432 @@
+                        copyNote = noteService.createNote(copyNote, document.getNoteTarget(),
+                                systemUser.getPrincipalId());
+                        Attachment copyAttachment = attachmentService.createAttachment(copyNote, outputFileName,
+                                ArConstants.TemplateUploadSystem.TEMPLATE_MIME_TYPE, copyReportStream.length,
+                                new ByteArrayInputStream(copyReportStream), KFSConstants.EMPTY_STRING);
+                        // adding attachment to the note
+                        copyNote.setAttachment(copyAttachment);
+                        noteService.save(copyNote);
+                        copyAttachment.setNoteIdentifier(copyNote.getNoteIdentifier());
+                        businessObjectService.save(copyAttachment);
+                        document.addNote(copyNote);
+                        invoiceAddressDetail.setNoteId(note.getNoteIdentifier());
+                        // saving the note to the document header
+                        documentService.updateDocument(document);
+                    } catch (IOException | DocumentException ex) {
+                        addNoteForInvoiceReportFail(document);
+                    }
+                } else {
+                    addNoteForInvoiceReportFail(document);
+                }
+            } else {
+                addNoteForInvoiceReportFail(document);
+            }
+
+        }
+    }
+
+    /**
+     * This method generated the template parameter list to populate the pdf invoices that are attached to the Document.
+     * The evident goal of this method was to return practically any possible property from the given document into a
+     * Map which could be stamped on any invoice PDF.  Given that, we've done some strange tricks with Map
+     * implementations.  First, we wrap the document in a ReflectionMap, which means that all nested properties from
+     * the document can be read via Map notation.  We still have a number of properties we want to add, though, for
+     * instance for the payee and the award.  So we wrap ReflectionMap in a FallbackMap, which will treat a regular
+     * HashMap and the ReflectionMap as if they were one Map for the sake of getting at least.  Finally, we wrap the map
+     * of all properties into a PdfFormattingMap, which formats any values to be returned through get() into properly
+     * formatted Strings.
+     *
+     * @param document the ContractsGrantsInvoiceDocument to convert into a Map form
+     * @return a Map.  With everything.
+     */
+    /*
+     * CU Customization: Use the version of this method from the 2019-10-31 financials patch,
+     * which properly lets the PdfFormattingMap handle converting the parameter values to Strings.
+     */
+    protected Map<String, String> getTemplateParameterList(ContractsGrantsInvoiceDocument document) {
```

Comment:
We already have an override of getTemplateParameterList() in our subclass to add some extra entries to the map, but I can look into updating our subclass so that we can do this as an override instead of an overlay. We still need the PdfFormattingMap overlay, though.

Reply:
perhaps moving this into our existing extension as getTemplateParameterListBase .... that way we can keep this change seperate from our customization, and still avoid an extra over lay.